pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,487 @@
+	// invariant since some other trace was registered and deregistered, but
+	// there is still a slot available. We will not register this trace if
+	// someone else is concurrently registering a trace on this store, but this
+	// is acceptable as it is a rare case.
+	registerSucceeded := r.numRegisteredStore.CompareAndSwap(numRegisteredStore, numRegisteredStore+1)
+	if registerSucceeded {
+		// Add one unconditionally to the replica count.
+		r.numRegisteredReplica.Add(1)
+	}
+	// Note we can't assert numRegisteredStore <= numAllowed because if the
+	// setting is changed it can be temporarily violated on other replicas.
+	return registerSucceeded
+}
+
+// tryStore attempts to store this value. If the index is already in the map it
+// will not store this entry and return false. It will also decrement counters
+// that were incremented by reserveSpace.
+// This is a rare case where we already have the index in the map. We
+// don't want to store this entry, but also need to decrement the
+// counter to avoid double tracing.
+func (r *RaftTracer) tryStore(tv *traceValue) *traceValue {
+	if existingTv, loaded := r.m.LoadOrStore(tv.traced.Index, tv); loaded {
+		r.numRegisteredReplica.Add(-1)
+		r.numRegisteredStore.Add(-1)
+		if tv.mu.propSpan != nil {
+			tv.logf(2, "duplicate registration ignored - existing trace: %s", existingTv)
+			tv.mu.propSpan.Finish()
+			// Don't allow additional tracing to this context.
+			tv.mu.propCtx = nil
+			tv.mu.propSpan = nil
+		}
+		existingTv.logf(2, "additional registration for same index: %s", tv)
+		return existingTv
+	}
+	return tv
+}
+
+// storeEntry stores the TracedEntry without a tracing context and returns the
+// newly created traceValue.
+func (r *RaftTracer) storeEntry(te kvserverpb.TracedEntry) *traceValue {
+	tv := traceValue{traced: te}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+	tv.mu.seenMsgAppResp = make(map[raftpb.PeerID]bool)
+	return r.tryStore(&tv)
+}
+
+// storeEntryWithTracing stores the TracedEntry along with a tracing context and
+// returns the newly created traceValue.
+func (r *RaftTracer) storeEntryWithTracing(
+	propCtx context.Context, span *tracing.Span, te kvserverpb.TracedEntry,
+) *traceValue {
+	tv := traceValue{traced: te}
+	tv.ctx = logtags.AddTag(r.ctx, "id", redact.Safe(tv.String()))
+	tv.mu.propCtx = propCtx
+	tv.mu.propSpan = span
+	tv.mu.seenMsgAppResp = make(map[raftpb.PeerID]bool)
+	return r.tryStore(&tv)
+}
+
+// RegisterRemote registers a remote trace. This is called when we receive a
+// raft message over the wire with a request to continue tracing it.
+func (r *RaftTracer) RegisterRemote(te kvserverpb.TracedEntry) {
+	if !r.reserveSpace() {
+		return
+	}
+	// NB: We don't currently return remote traces, if we did, we would pass the
+	// remote ctx here and trace it. The problem is knowing when to send it
+	// back to the remote node.
+	tv := r.storeEntry(te)
+	tv.logf(1, "registering remote trace %s", tv)
```

Comment:
Probably don't want this logging message in case `storeEntry` returns a previously existing trace.

Reply:
Good point. I cleaned this up to avoid the dupe log.